### DISCUSSION TITLE
Deleted messages not being shown

### DISCUSSION THREAD

**Discussion by albionb96** [nats-io/nats-server#5096] [Q&A] [answered]
After starting multiple producers/consumers which publish with different subjects on the same jetstream I do not see the number of deleted messages in nats-box. As you can see from the following screenshot the stream size is ca. 600k messages, which is ok as expected and also the number of total sent messages is more than 43,6Mil which is also correct. But I would expect too se under deleted messages not 0 but something around 43Mil, which I don't see. When I started another java project with multiple producers/consumers I was able to see the number of deleted messages but this time not any more.

Should this parameter in nats-box show the number of deleted messages or am I wrong, because I couldn't find something in this regard?

![image](https://github.com/nats-io/nats-server/assets/49485798/da75efab-70ba-4146-98b5-4a3d9d3f1eb0)

**Comment by ripienaar** [comment, marked as answer]
Imagine you have a stream with messages 1,2,3,4,5,6,7,8,9,10 the stream removes messages due to limits so you have 5,6,7,8,9,10 this would not be reported as deleted messages.

If however you now delete using the delete API message 8, you would have 1 deleted message.  If you deleted message 5 you would still have 1 deleted message reported.

So its about the definition of deleted messages - they are gaps in the stream if you had 5,6,7,|gap|,10 the gap is 2 big so deleted is 2.

**Reply by albionb96** [reply]
@ripienaar thank you very much!

**Reply by ripienaar** [reply]
The same pattern would be seen if you say keep 5 messages per subject, if one subject is busy getting messages - so always removing messages - and other subjects dont change, this will cause a lot of gaps and so lots of deleted messages

**Reply by stefanLeo** [reply]
So just to verify if I understood that correct:
If I have a stream with 2 subjects and a limit of 3.
I have 1,2,3,4,5,6 in the steam.
Subject1: 1,4,5
Subject2: 2,3,6
Now someone adds a msg to subject2, then I would have 1 deleted message as #2 would get deleted due to limits right? And this would add up as long as  stream 1 receives no new messages..

I.e., deleted counts gaps on a stream level across subjects and tells you something about "fragmentation" of the stream. Is it a bad thing (performance wise)  to have high deleted numbers?

**Reply by derekcollison** [reply]
Correct. Deleted msgs are always in the interior, between first and last sequences.

2.10 is much better at handling large interior delete state.

**Reply by albionb96** [reply]
@derekcollison we came around deleted messages actually by accident, when looking at the nats-box. It would be great to have a prom metric for the deleted messages as well, and maybe incorporate it in the nats grafana dashboards.

**Reply by derekcollison** [reply]
Looping in @Jarema

**Reply by sourabhaggrawal** [reply]
I'm trying to understand the exact value that 'Deleted' represents, as I've consistently observed it being 0. Can we consider the difference between the stream's last sequence and the messages available in the stream to determine the exact number of messages deleted? If that’s the intended purpose, please let me know if I'm missing something."

**Reply by ripienaar** [reply]
If a stream had messages 1.3.4 because you used the APi to deleted message 2 you would have one deleted. 

If it had 2.3.4 you have no deleted. 

The deleted tells you the size of the gaps.

**Reply by sourabhaggrawal** [reply]
with the API you mean explicitly call to Delete Msg API ?  also what does the gap means, is it lost messages ?

**Reply by ripienaar** [reply]
Yes, well anything that deletes messages. Could be a WQ ack, could be limits apply to subjects etc. all will cause it.

**Reply by sourabhaggrawal** [reply]
Yes thats what I anticipated and would expect it to show some numbers in Deleted Message but I have seen this Deleted as 0 every time in stream report.
I have tested this by publishing and consuming 1000 messages, deleted is still shown as 0 also for some reason First Sequence is updated to 1001 where I have only published 1000 messages.
<img width="535" alt="Screenshot 2024-08-30 at 12 06 18 AM" src="https://github.com/user-attachments/assets/1a7999ba-b511-4060-a7b5-17f1fafb94c3">

**Reply by sourabhaggrawal** [reply]
I am using nats-server 2.10.18 and nats-0.1.5 cli

**Reply by ripienaar** [reply]
If you consumed all first 1000 you wi t see any deleted as there will be no gap

**Reply by neilalexander** [reply]
The deletes are showing how many messages were deleted _between the first and last sequence_. If your delete caused the first or last sequence to change because it was the first message, it won't be counted as a delete within that range, because the range itself changed.

**Reply by ripienaar** [reply]
```
$ nats s add DELETED --subjects 'js.in.deleted.>' --defaults
$ for i in 1 2 3 4 5
do
nats req 'js.in.deleted.1' '' --count 100
nats req 'js.in.deleted.2' '' --count 100
done
$ nats s purge DELETED --subject js.in.deleted.1
```

So we make a stream, put in messages such that they are in groups such that when I purge the js.in.deleted.1 subject I have gaps, now stream info showsL:

```
State:

              Messages: 500
                 Bytes: 22 KiB
        First Sequence: 101 @ 2024-08-29 19:20:27
         Last Sequence: 1,000 @ 2024-08-29 19:20:33
      Deleted Messages: 400
      Active Consumers: 0
    Number of Subjects: 1
```

And we can see the gaps:

```
$ nats s gaps DELETED -f
500 / 500 [====================================================================] 100%

╭──────────────────────────────╮
│ 4 gaps found in Stream DELET │
│              ED              │
├───────────────┬──────────────┤
│ First Message │ Last Message │
├───────────────┼──────────────┤
│ 201           │ 300          │
│ 401           │ 500          │
│ 601           │ 700          │
│ 801           │ 900          │
╰───────────────┴──────────────╯
```

**Reply by sourabhaggrawal** [reply]
one of my stream is WQ with only one subject, no TTL & only one consumer reads from it, no explicit call to delete api. 
Now that there is only one consumer reading from it, I am hoping it will read sequentially and nats server keep deleting the messages when ack'd and FirstSequence should be moving forward accordingly. 
Yet I see FirstSequence is left far behind and only few messages are left in stream. I have not deleted/purged message or stream or subject anywhere. I still don't  understand why it is showing Deleted now. However there are some redelivery of messages that might have happened
![Screenshot 2024-08-30 at 12 20 46 PM](https://github.com/user-attachments/assets/dd44b037-d830-4e42-adfb-c9dd42dbc194)

**Reply by ripienaar** [reply]
Probably some was not asked before max delivered.

**Comment by sourabhaggrawal** [comment]
Also when I run gaps command on stream, it throws me this error

```
? Really detect gaps in stream STREAM_1 with 277 messages and 1.1 MiB bytes Yes
panic: runtime error: index out of range [0] with length 0

goroutine 50 [running]:
github.com/gosuri/uiprogress.(*Bar).Bytes(0xc000312300)
	/go/amd64/pkg/mod/github.com/gosuri/uiprogress@v0.0.1/bar.go:195 +0x40b
github.com/gosuri/uiprogress.(*Bar).String(...)
	/go/amd64/pkg/mod/github.com/gosuri/uiprogress@v0.0.1/bar.go:214
github.com/gosuri/uiprogress.(*Progress).print(0xc00013e420)
	/go/amd64/pkg/mod/github.com/gosuri/uiprogress@v0.0.1/progress.go:127 +0x8f
github.com/gosuri/uiprogress.(*Progress).Listen(0xc00013e420)
	/go/amd64/pkg/mod/github.com/gosuri/uiprogress@v0.0.1/progress.go:114 +0x25
created by github.com/gosuri/uiprogress.(*Progress).Start in goroutine 1
	/go/amd64/pkg/mod/github.com/gosuri/uiprogress@v0.0.1/progress.go:134 +0x4f
```

**Reply by ripienaar** [reply]
I think that's fixed in the main branch of `nats` - which is what I used.

Get latest with `curl -sf https://binaries.nats.dev/nats-io/natscli/nats@main|bash`

**Reply by sourabhaggrawal** [reply]
I think I get this now, how this works. Tried to simulate this 10 different times to validate the expectation.
But where we can use this Deleted or Gap ?

**Reply by ripienaar** [reply]
its not very useful to be honest, if the deleted counts are huge you might have some higher memory use etc, and the gap detection could help you with commands like `nats s purge` to purge up to a gap which will then drop the deleted messages down.

**Reply by sourabhaggrawal** [reply]
Yeah, with delete I was hoping it will give me the exact number of messages that has been read and removed from stream which could be used actually. 
I guess changing the logic to consider (Last Sequence - Available messages) will give you the exact number, no ?

**Reply by ripienaar** [reply]
yup

**Reply by sourabhaggrawal** [reply]
> its not very useful to be honest, if the deleted counts are huge you might have some higher memory use etc, and the gap detection could help you with commands like `nats s purge` to purge up to a gap which will then drop the deleted messages down.

So I have to make sure TTL is configured with every stream to delete the messages which are left in stream after redelivery exhausted. Is there any metric that can be used to configure alert to notify about this, any plan for that ?

**Reply by sourabhaggrawal** [reply]
> yup
Shall I raise a PR for this ?

**Reply by ripienaar** [reply]
Pr to show total deleted messages? Not sure its useful enough to add to the server but its something you can know how to figure out from nats cli output maybe

**Reply by neilalexander** [reply]
If nothing else we might want to update the NATS CLI to call it `Interior Deletes` instead of `Deleted Messages` to avoid confusion.

**Reply by sourabhaggrawal** [reply]
Sure, I can handle the calculation, and I think fixing it permanently will benefit others as well, saving them from the frustration 😄  . Do you agree?

**Reply by sourabhaggrawal** [reply]
> If nothing else we might want to update the NATS CLI to call it `Interior Deletes` instead of `Deleted Messages` to avoid confusion.

But I am not sure what you mean by Interior Deletes and if everyone will get it right.

**Reply by ripienaar** [reply]
I don't think its broken, its just not showing what you thought?

@neilalexander personally never liked "Interior Deletes" never seemed to explain it to me :)

**Reply by sourabhaggrawal** [reply]
> I don't think its broken, its just not showing what you thought?
> 
> @neilalexander personally never liked "Interior Deletes" never seemed to explain it to me :)

Yeah I mean changing the logic little bit will show you what exactly it is supposed to, I have tried it locally and looks fine to me.

**Reply by ripienaar** [reply]
It's showing what it is supposed to :)

**Reply by sourabhaggrawal** [reply]
It's confusing and difficult to explain it to somebody too, also sometime seeing it as some value and sometime 0, its difficult to digest actually :P.
With Deleted, folks would expect it to show total number of messages that stream had receive from beginning and how many have been deleted.

**Reply by neilalexander** [reply]
Showing the interior deletes is useful for us because interior deletes are one of the first things we look for when there are performance issues on a stream. It probably makes sense for it to be named as such — it might not be universally descriptive, but at least it's not misleading in the way `Deleted Messages` is.

I can understand the desire to show the overall count of deletes too, but keep in mind that the first sequence of a stream isn't guaranteed to be 1, it can be set to whatever you want when the stream is created, so calculating from last seq and available messages alone isn't necessarily going to do the right thing.

**Reply by sourabhaggrawal** [reply]
> Showing the interior deletes is useful for us because interior deletes are one of the first things we look for when there are performance issues on a stream. It probably makes sense for it to be named as such — it might not be universally descriptive, but at least it's not misleading in the way `Deleted Messages` is.

I agree with you Deleted Message in misleading and something to be done with it.

> 
> I can understand the desire to show the overall count of deletes too, but keep in mind that the first sequence of a stream isn't guaranteed to be 1, it can be set to whatever you want when the stream is created, so calculating from last seq and available messages alone isn't necessarily going to do the right thing.

I did not understand this part, can you set the number of sequence to start with while creating the stream ? Are you referring to source and mirror stream here ?

**Reply by neilalexander** [reply]
>can you set the number of sequence to start with while creating the stream ?

Yes:
```
$ nats stream create test --defaults --subjects test --first-sequence=100
Stream test was created

Information for Stream test created 2024-08-30 15:07:02

              Subjects: test
              Replicas: 1
               Storage: File
        First Sequence: 100

Options:

             Retention: Limits
       Acknowledgments: true
        Discard Policy: Old
      Duplicate Window: 2m0s
            Direct Get: true
     Allows Msg Delete: true
          Allows Purge: true
        Allows Rollups: false

Limits:

      Maximum Messages: unlimited
   Maximum Per Subject: unlimited
         Maximum Bytes: unlimited
           Maximum Age: unlimited
  Maximum Message Size: unlimited
     Maximum Consumers: unlimited

State:

              Messages: 0
                 Bytes: 0 B
        First Sequence: 100
         Last Sequence: 99
      Active Consumers: 0
```

**Reply by ripienaar** [reply]
to be fair setting the start seq is a special case, but this really just isn't that difficult to figure out deleted messages.

```
           Messages: 1,000
      Last Sequence: 8,689,458 @ 2024-08-30 14:13:06
```

so like 8,689,458 - 1000 messages were removed.

Seems to be a huge deal over simple math to me

**Reply by sourabhaggrawal** [reply]
> > can you set the number of sequence to start with while creating the stream ?
> 
> Yes:
> 
> ```
> $ nats stream create test --defaults --subjects test --first-sequence=100
> Stream test was created
> 
> Information for Stream test created 2024-08-30 15:07:02
> 
>               Subjects: test
>               Replicas: 1
>                Storage: File
>         First Sequence: 100
> 
> Options:
> 
>              Retention: Limits
>        Acknowledgments: true
>         Discard Policy: Old
>       Duplicate Window: 2m0s
>             Direct Get: true
>      Allows Msg Delete: true
>           Allows Purge: true
>         Allows Rollups: false
> 
> Limits:
> 
>       Maximum Messages: unlimited
>    Maximum Per Subject: unlimited
>          Maximum Bytes: unlimited
>            Maximum Age: unlimited
>   Maximum Message Size: unlimited
>      Maximum Consumers: unlimited
> 
> State:
> 
>               Messages: 0
>                  Bytes: 0 B
>         First Sequence: 100
>          Last Sequence: 99
>       Active Consumers: 0
> ```

This is interesting, what is the use case of this ?
